darts/utils/data/encoder_base.py

```diff
@@ -0,0 +1,382 @@
+        if covariate is not None:
+            return covariate.time_index
+        else:
+            return _generate_index(start=target.end_time() - target.freq * (self.input_chunk_length - 1),
+                                   length=self.input_chunk_length + max(0, n - self.output_chunk_length),
+                                   freq=target.freq)
+
+
+class FutureCovariateIndexGenerator(CovariateIndexGenerator):
+    """Generates index for future covariates on train and inference datasets."""
+    def generate_train_series(self,
+                              target: TimeSeries,
+                              covariate: Optional[TimeSeries] = None) -> SupportedIndexes:
+        """For training (when `n` is `None`) we can simply use the future covariates (if available) or target as
+        reference to extract the time index.
+        """
+
+        super(FutureCovariateIndexGenerator, self).generate_train_series(target, covariate)
+
+        # save a reference index if specified
+        if self.reference_index_type is not ReferenceIndexType.NONE and self.reference_index is None:
+            if self.reference_index_type is ReferenceIndexType.PREDICTION:
+                self.reference_index = (len(target) - 1, target.end_time())
+            else:  # save the time step before start of target series
+                self.reference_index = (-1, target.start_time() - target.freq)
+
+        return covariate.time_index if covariate is not None else target.time_index
+
+    def generate_inference_series(self,
+                                  n: int,
+                                  target: TimeSeries,
+                                  covariate: Optional[TimeSeries] = None) -> SupportedIndexes:
+        """For prediction (`n` is given) with future covariates we have to distinguish between two cases:
+        1)  If future covariates are given, we can use them as reference
+        2)  If future covariates are missing, we need to generate a time index that starts `input_chunk_length`
+            before the end of `target` and ends `max(n, output_chunk_length)` after the end of `target`
+        """
+        super(FutureCovariateIndexGenerator, self).generate_inference_series(n, target, covariate)
+
+        if covariate is not None:
+            return covariate.time_index
+        else:
+            return _generate_index(start=target.end_time() - target.freq * (self.input_chunk_length - 1),
+                                   length=self.input_chunk_length + max(n, self.output_chunk_length),
+                                   freq=target.freq)
+
+
+class Encoder(ABC):
+    """Abstract class for all encoders"""
+
+    @abstractmethod
+    def __init__(self):
+        self.attribute = None
+        self.dtype = np.float64
+
+    @abstractmethod
+    def encode_train(self,
+                     target: TimeSeries,
+                     covariate: Optional[TimeSeries] = None,
+                     merge_covariate: bool = True,
```

Comment:
yes, the SequenceEncoder sets merge_covariate to False when it calls the SingleEncoders so that merging the encodings with optional user-supplied covariates is only applied once when all encodings were generated.